beacon_node/eth1/src/cache.rs

```diff
@@ -0,0 +1,182 @@
+            // it waits for other spawned futures to complete on current thread.
+            if let Ok((block_number, eth1_data)) = tokio::runtime::current_thread::block_on_all(
+                fetch_eth1_data(distance, current_block_number, self.fetcher.clone()),
+            )? {
+                let mut cache_write = self.cache.write();
+                cache_write.insert(block_number, eth1_data);
+                return Ok(cache_write
+                    .get(&block_number)
+                    .ok_or(Error::InvalidData)? // Note: Is this the right error type?
+                    .clone());
+            }
+        }
+        Err(Error::InvalidData)
+    }
+
+    /// Returns a Vec<Eth1Data> corresponding to given distance range.
+    pub fn get_eth1_data_in_range(&self, start: u64, end: u64) -> Vec<Eth1Data> {
+        (start..end)
+            .map(|h| self.get_eth1_data(h))
+            .flatten() // Chuck Err values. This might be okay since its unlikely that the entire range returns None.
+            .collect::<Vec<Eth1Data>>()
+    }
+}
+
+fn fetch_eth1_data_in_range<F: Eth1DataFetcher>(
+    start: u64,
+    end: u64,
+    current_block_number: U256,
+    fetcher: Arc<F>,
+) -> impl Stream<Item = Result<(U256, Eth1Data)>, Error = Error> + Send {
+    stream::futures_ordered(
+        (start..end).map(move |i| fetch_eth1_data(i, current_block_number, fetcher.clone())),
+    )
+}
+
+/// Fetches Eth1 data from the Eth1Data fetcher object.
+fn fetch_eth1_data<F: Eth1DataFetcher>(
+    distance: u64,
+    current_block_number: U256,
+    fetcher: Arc<F>,
+) -> impl Future<Item = Result<(U256, Eth1Data)>, Error = Error> + Send {
+    let block_number: U256 = current_block_number
+        .checked_sub(distance.into())
+        .unwrap_or(U256::zero());
+    let deposit_root = fetcher.get_deposit_root(Some(BlockNumber::Number(block_number.as_u64())));
+    let deposit_count = fetcher.get_deposit_count(Some(BlockNumber::Number(block_number.as_u64())));
+    let block_hash = fetcher.get_block_hash_by_height(block_number.as_u64());
+    let eth1_data_future = deposit_root.join3(deposit_count, block_hash);
+    eth1_data_future.map(move |data| {
+        let eth1_data = Eth1Data {
+            deposit_root: data.0,
+            deposit_count: data.1?,
+            block_hash: data.2.ok_or(Error::DecodingError)?,
+        };
+        Ok((block_number, eth1_data))
+    })
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::web3_fetcher::Web3DataFetcher;
+    use std::time::{Duration, Instant};
+    use tokio::timer::{Delay, Interval};
+
+    // Note: Running tests using ganache-cli instance with config
+    // from https://github.com/ChainSafe/lodestar#starting-private-eth1-chain
+
+    fn setup() -> Web3DataFetcher {
+        let deposit_contract_address = "8c594691C0E592FFA21F153a16aE41db5beFcaaa";
+        let mut abi_path = std::env::current_dir().unwrap();
+        abi_path.push("deposit_contract.json");
```

Comment:
I think we should include this ABI in the compiled binary using [`include_bytes`](https://doc.rust-lang.org/std/macro.include_bytes.html). The file will need to be present at compile time, however it will not be required at runtime.

I think we should add the following [`build.rs`](https://gist.github.com/paulhauner/d12168f9f03639af40565cadd0ee78f2) build script to the `eth1` crate. It downloads the ABI from the eth2.0-specs repo. _Note the "this is broken" TODO._

Reply:
Fixed the `build` script.

Reply:
Cargo has a built-in concept of "build scripts". They are run before compilation (e.g., during `cargo build`). Docs: https://doc.rust-lang.org/cargo/reference/build-scripts.html

Basically, move `eth1/src/bin/build_abi.rs` to `eth1/build.rs` and add `build = "build.rs"` to `eth1/Cargo.toml` :)

Reply:
Done. This is cool stuff!